Fix: regex fails to capture certain langs

Qrbaker opened the pull request:
Small fix that I noticed when testing various languages today:


Languages with non alphanumeric characters (i.e. c++)
would not be captured by the regex for syntax highlighting.

This change moves from a positive capture (all alphanumeric
characters) to a negative capture (anything that is NOT a ").

This should be a more robust solution... at least until some-
one invents a language with a " in the name.

Qrbaker committed:
Fix: regex fails to capture certain langs

Languages with non alphanumeric characters (i.e. c++)
would not be captured by the regex for syntax highlighting.

This change moves from a positive capture (all alphanumeric
characters) to a negative capture (anything that is NOT a ").

This should be a more robust solution... at least until some-
one invents a language with a " in the name.

TrueBrain commented on wikitexthtml/render/preprocess.py:
```diff
@@ -18,7 +18,7 @@
 PRE_PATTERN = regex.compile(r"<pre>([\s\S]*?)(?></pre>|\Z)")
 SYNTAXHIGHLIGHT_PATTERN = regex.compile(
-    r"<syntaxhighlight( lang=\")?(?P<lang>[a-zA-z]+)?\"?"
+    r"<syntaxhighlight( lang=\")?(?P<lang>[^\"]+)?\"?"
```
While at it, there is something else strange about this regex .. take the following example:

`<syntaxhighlight c++ line 12>`

This is selecting the language `c++ line 12`. On the other hand:

`<syntaxhighlight c++" line 12>`

Does the same as:

`<syntaxhighlight lang="c++" line 12>`

A possible solution is to drop `lang=` completely, or otherwise make it an option: either `lang="c++"` or `c++`. Or don't support `c++` (without `lang`) syntax at all.

Not sure if you picked an existing syntax for enabling the highlighter or that you made this syntax up, but looking at how most (not all) of mediawiki works, one could also opt for having `lang="c++" line=12` as only valid syntax.

What do you think?

Qrbaker replied:
I'm just trying to replicate mediawiki syntax for `lang` and `line`. The example `<syntaxhighlight c++ line 12>` is undefined behavior as far as I can tell from https://www.mediawiki.org/wiki/Extension:SyntaxHighlight#Parameters

TrueBrain replied:
Ah, so it is existing! And `start=` got left out ;) Let me see if I can help a bit with this regex ... I hate regex :P

Qrbaker replied:
I can add it, if you think it should be in there.

TrueBrain replied:
I think we should either mimic what mediawiki is doing, or do our own thing. Being "something in between" would confuse the users, I think? (open for opinions here :D)

TrueBrain replied:
Something like this:
`<syntaxhighlight( lang=\"(?P<lang>[^\"]+)\")?(?P<line> line( start=\"(?P<lineStart>\d+)\")?)?>`
Maybe?

TrueBrain replied:
(only difference is that I made `line=""` part of a group, so you cannot have `lang` captured without having `lang=""` there. And use `start=` instead of just a number ;))

Qrbaker replied:
https://regex101.com/r/aojXe7/1 < I think this should handle everything now. Happy for suggestions.

TrueBrain replied:
I suggest to take a look at my regex above ;) Yours still allows for `<syntaxhighlight java" line start=12>blashf</syntaxhighlight>` as input, which I would consider buggy ;)

Or maybe put differently: why didn't you capture `lang` together with the `lang=""` part, but why use a 3-part which all 3 parts optional?

TrueBrain replied:
In a nice linky: https://regex101.com/r/UPfWCA/1

Qrbaker replied:
> why didn't you capture lang together with the lang="" part, but why use a 3-part which all 3 parts optional?

Because I don't do regex and built this linearly over the course of months, that's why 🙃

TrueBrain replied:
Always nice to know I am not alone :P Regex are the worst .. reason I am so happy with https://github.com/5j9/wikitextparser .. he did all the complicated regex for us :D

If you like what you see in my link, I suggest to just use my regex. But feel free to fiddle with it yourself .. it is an endless battle :P

Qrbaker replied:
Yeah works for me, I pushed that along with another tweak to the render() method.

Qrbaker committed:
Fix: Tweaked regex to be more robust

Also adjusted the render() method to handle the arguments
expected from downstream overrides
